src/Bicep.Core/Semantics/Namespaces/DefaultNamespaceProvider.cs

```diff
@@ -5,29 +5,55 @@
 using System.Collections.Immutable;
 using Bicep.Core.Extensions;
 using Bicep.Core.Features;
+using Bicep.Core.Syntax;
 using Bicep.Core.TypeSystem;
 using Bicep.Core.TypeSystem.Az;
 
 namespace Bicep.Core.Semantics.Namespaces;
 
 public class DefaultNamespaceProvider : INamespaceProvider
 {
-    private delegate NamespaceType GetNamespaceDelegate(string aliasName, ResourceScope resourceScope, IFeatureProvider features);
+    private delegate NamespaceType GetNamespaceDelegate(
+        string aliasName,
+        ResourceScope resourceScope,
+        IFeatureProvider features,
+        ImportDeclarationSyntax? importDeclarationSyntax = null);
     private readonly ImmutableDictionary<string, GetNamespaceDelegate> providerLookup;
+    private readonly IAzResourceTypeLoaderFactory azResourceTypeLoaderFactory;
 
-    public DefaultNamespaceProvider(IAzResourceTypeLoader azResourceTypeLoader)
+    public DefaultNamespaceProvider(IAzResourceTypeLoaderFactory azResourceTypeLoaderFactory)
     {
-        var azResourceTypeProvider = new AzResourceTypeProvider(azResourceTypeLoader);
+        this.azResourceTypeLoaderFactory = azResourceTypeLoaderFactory;
         this.providerLookup = new Dictionary<string, GetNamespaceDelegate>
         {
-            [SystemNamespaceType.BuiltInName] = (alias, scope, features) => SystemNamespaceType.Create(alias, features),
-            [AzNamespaceType.BuiltInName] = (alias, scope, features) => AzNamespaceType.Create(alias, scope, azResourceTypeProvider),
-            [K8sNamespaceType.BuiltInName] = (alias, scope, features) => K8sNamespaceType.Create(alias),
+            [SystemNamespaceType.BuiltInName] = (alias, scope, features, ids) => SystemNamespaceType.Create(alias, features),
+            [AzNamespaceType.BuiltInName] = (alias, scope, features, ids) =>
+            {
+                var loader = azResourceTypeLoaderFactory.GetResourceTypeLoader(ids, features);
+                if (loader is null)
+                {
+                    return null!;
```

Comment:
It would be better to change the return type of `GetNamespaceDelegate` to nullable and avoid using non-null assertions.

Reply:
I don't understand how making the return type of GetNamespaceDelegate to nullable would help here, can you please elaborate?

Reply:
Except maybe in unit tests, you should never use null!.  It lets you return null even though you're telling the compiler it's not null.  

I think it's actually the Dictionary<string, GetNamespaceDelegate> that you're abusing null! return.  If azResourceTypeLoaderFactory.GetResourceTypeLoader() should never return null, then you should throw an error instead.  Otherwise it should be Dictionary<string, GetNamespaceDelegate?>.

Reply:
I think what needs to happen here is the return value of `GetNamespaceDelegate` should be nullable. Refactoring like this and resolving. Feel free to open up again if that is not a good solution